File: Maps/SortRelativeArray/sort.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;


class Solution {
public:
    void relativeSort(vector<int>& arr1, vector<int>& arr2) {
        unordered_map<int, int> mp;

        int n2 = arr2.size();
        for (int i = 0; i < n2; i++) {
            mp[arr2[i]] = mp[arr2[i]];
        }

        int n1 = arr1.size();
        for (int i = 0; i < n1; i++) {
            mp[arr1[i]]++;
        }

        vector<int> ans;
        vector<int> remEle;

        for (int i = 0; i < n2; i++) {
            if (mp.find(arr2[i]) != mp.end()) {
                remEle.insert(remEle.end(), mp[arr2[i]], arr2[i]);
                mp.erase(arr2[i]);
            }
        }

        vector<pair<int, int>> temp(mp.begin(), mp.end());
        sort(temp.begin(), temp.end());

        for (auto &ele : temp) {
            remEle.insert(remEle.end(), ele.second, ele.first);
        }

        ans.insert(ans.end(), remEle.begin(), remEle.end());

        arr1 = ans;
    }
};

File: Maps/SortRelativeArray/sort.py
class Solution:
    def relativeSort(self, arr1, arr2):
        mp = {}
        n2 = len(arr2)
        for i in range(n2):
            mp[arr2[i]] = mp.get(arr2[i],0)
        n1 = len(arr1)
        for i in range(n1):
            mp[arr1[i]] = mp.get(arr1[i],0) +1
        ans = []
        remEle = []
        
        for i in range(n2):
            if arr2[i] in mp:
                ans.extend([arr2[i]]*mp[arr2[i]])
                del mp[arr2[i]]
            
        for key in sorted(mp):
            remEle.extend([key] * mp[key])

        ans.extend(remEle)
        arr1[:] = ans 
        
        return arr1

